mirrord/cli/src/port_forward.rs

```diff
@@ -0,0 +1,689 @@
+                                    }
+                                },
+                                None => break Ok(()), // PortForwarder received DaemonTcpOutgoing::Close
+                            }
+                        }
+                    }
+                });
+            }
+            (socket, Err(error)) => {
+                // error from TcpStream
+                tracing::error!("Error occured while listening to local socket {socket}: {error}");
+                self.listeners.remove(&socket);
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_msg_from_task(
+        &mut self,
+        message: PortForwardMessage,
+    ) -> Result<(), PortForwardError> {
+        match message {
+            PortForwardMessage::Connect(remote_socket, oneshot) => {
+                let remote_address = SocketAddress::Ip(remote_socket);
+                self.oneshots.push_back(oneshot);
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Connect(
+                        LayerConnect { remote_address },
+                    )))
+                    .await
+                else {
+                    return Err(PortForwardError::AgentError(
+                        "Failed to send ClientMessage to agent".into(),
+                    ));
+                };
+            }
+            PortForwardMessage::Send(connection_id, bytes) => {
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Write(
+                        LayerWrite {
+                            connection_id,
+                            bytes,
+                        },
+                    )))
+                    .await
+                else {
+                    return Err(PortForwardError::AgentError(
+                        "Failed to send ClientMessage to agent".into(),
+                    ));
+                };
+            }
+            PortForwardMessage::Close(socket, connection_id) => {
+                // end of listener TCP stream in task
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Close(
+                        LayerClose { connection_id },
+                    )))
+                    .await
+                else {
+                    return Err(PortForwardError::AgentError(
+                        "Failed to send ClientMessage to agent".into(),
+                    ));
+                };
+                self.tasks.remove(&socket);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum PortForwardError {
```

Comment:
Error messages here should probably start with a lowercase letter (this probably is not always the case across the code base, but nvm) to make the final user-facing message look better. Same for nested messages